### DISCUSSION TITLE
Using CURL to connect via HA (MQTT service or light service)

### DISCUSSION THREAD

**Discussion by erwin** [HASwitchPlate/openHASP#378] [Show and tell]
Here's a brief reference for using `curl` to send commands to your OpenHASP devices if using Home Assistant.

First you need a `Long-Lived Access Token` from Home-Assistant.

Log in, click your own "Profile", scroll to the bottom and choose "Create Token"

![image](https://user-images.githubusercontent.com/11534217/187825995-0922143b-88f9-4633-9cea-6e672df4b3ba.png)

![image](https://user-images.githubusercontent.com/11534217/187826073-fa4710a1-d01d-4173-a1f9-80404c53184d.png)

Then you'll get a token like this:
![image](https://user-images.githubusercontent.com/11534217/187826138-c53db252-d133-4ab4-87a9-5bb3eef2715c.png)

(don't worry, I only generated that for these screenshots and have already deleted that token in my system).

To directly set values via the Home-Assistant `/api/services/light/turn_on`

You'll need to start your script with something like:

(of course your value for BEARER has to be something that you generated on your home-assistant)

You also need to set HA_URL to the URL you actually use at home to connect to your home-assistant.

```
#!/bin/bash

export BEARER=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiIyYTliMjgwZDIyMTI0ODIzODU0ZDYxY2ViYjQ3NDBmNiIsImlhdCI6MTY2MjAwMzE3NywiZXhwIjoxOTc3MzYzMTc3fQ.94g5wA8OwOOSHjibD1hIqhpmcHjH5I9rIeGrlgVmdsI

export HA_URL=http://192.home.assistant.240:8123
```

Then you can define a bash function like:

```
function MOODLIGHT() {
  curl -X POST \
    -H "Authorization: Bearer $BEARER" \
    -H 'Content-Type: application/json' \
    -d "{
        \"entity_id\": \"light.$1_moodlight\",
        \"rgb_color\": [$2, $3, $4],
        \"brightness_pct\": $5
        }" \
     "${HA_URL}/api/services/light/turn_on"
}
```

For example, to set the lights to RED, you would just use:

```
MOODLIGHT YOUR_PLATE_NAME 255     0    0          100
#                         RED GREEN BLUE BRIGHTNESS %

# moodlight to yellow, 100% brightness
MOODLIGHT YOUR_PLATE_NAME 255 255 0 100

# moodlight to cyan, 100% brightness
MOODLIGHT YOUR_PLATE_NAME 0 255 255 100
```

You just use a Red - Green - Blue value from 0 - 255

The home-assistant services that I use the most this way are:
* `"${HA_URL}/api/services/light/turn_on"`
* `"${HA_URL}/api/services/light/turn_off"`
* `"${HA_URL}/api/services/light/toggle"`
* `"${HA_URL}/api/services/switch/turn_on"`
* `"${HA_URL}/api/services/switch/turn_off"`
* `"${HA_URL}/api/services/switch/toggle"`
* `"${HA_URL}/api/services/homeassistant/turn_on"`
* `"${HA_URL}/api/services/homeassistant/turn_off"`

The latter two I mostly use for `input_boolean` values that I display on the OpenHASP screens.

Home-Assistant ultimately forwards commands to OpenHASP via MQTT. It's also possible to send directly to MQTT via Home Assistant, though I have not benchmarked to see if the total system response time is faster or slower.

To send a request via curl throught the HA MQTT service, you can use:

```
function BACKLIGHT() {
  curl --silent -X POST \
    -H "Authorization: Bearer $BEARER" \
    -H "Content-Type: application/json" \
    -d "{
          \"payload\": \"backlight {'state': $3, 'brightness': $2}\",
          \"topic\": \"hasp/$1/command\"
        }" \
        "${HA_URL}/api/services/mqtt/publish"
}
```

You would just run:

```
# turn the plate on, medium brightness
BACKLIGHT MY_PLATE_NAME 128 true

# turn the plate on, max brightness
BACKLIGHT MY_PLATE_NAME 255 true

# turn the plate off
BACKLIGHT MY_PLATE_NAME 0 false
```

If you're not very familiar with shell script, I would just create a new `openhasp.sh` file and paste the above in... On your command line just run:

```
source ./openhasp.sh
```

Then try to call the functions.

If you get an error, just edit the `openhasp.sh` file and run `source ./openhasp.sh` to update it and try again.

Hopefully this helps you have more for on the command line and in your shell scripts connecting things to OpenHASP!

**Comment by nagyrobi** [comment]
Thanks for sharing!
